### ISSUE TITLE
[GiDoutput][ShapeOpt] Error when postprocessing intermediate results

### ISSUE THREAD

**Issue by armingeiser** [KratosMultiphysics/Kratos#3505]
I currently have a problem when i try to postprocess my case in GiD while Kratos is still running.

```
Some errors found in the results file:
Error reading file '/Optimization_Results/design_history.post.bin': Vector result 'DF1DX' of analysis 'Kratos', step '1' : index greater than last node: 1000001 > 241523.error reading values for Vector result 'DF1DX' of analysis 'Kratos', step '1' :  Only 241523 out of 244966 values read..
```

In GiD i can visualize the geometry, but the results are not loaded.

However, after the whole Kratos run is done, the result file can suddenly be postprocessed.

This happens to me only with a rather large example, that i cannot share unfortunately. If someone has a hint i would really appreciate it.

**Comment by armingeiser** [comment]
With smaller cases, the error is similar but not the same:

```
Some errors found in the results file:
Error reading file '/Optimization_Results/design_history.post.bin': error reading values for Vector result 'SHAPE_CHANGE' of analysis 'Kratos', step '4' : <fd:22>: unexpected end of file Only 1369 out of 2905 values read.<fd:22>: unexpected end of file
<fd:22>: unexpected end of file.
```

Here the error always happens in the last result of the latest step, still annoying but at least all results that have been imported by GiD until then can be postprocessed.

**Comment by pooyan-dadvand** [comment]
you may flush the io to make it write the cached buffer. But this can reduce the performance in clusters in which writing to disk is very slow...

**Comment by armingeiser** [comment]
I tried to call the `GidIO::Flush()` function after the `PrintOutput` of the gid_output_process.
It helps for the `Ascii` mode, for binary results, the file still seems not to be written to disk completely.

**Comment by pooyan-dadvand** [comment]
@maceligueta any hint?

**Comment by maceligueta** [comment]
I've noticed something similar for cases where the output is in Single file mode, not in Multiple. I did not dig deep in that, but I suspect we should close the file and open it again for appending new results... If it's kept open, maybe that's the reason why Gid does not want to open it. Not really sure about this.

**Comment by maceligueta** [comment]
Note that this line closes the results only for Ascii :    
https://github.com/KratosMultiphysics/Kratos/blob/55e7608139955207e0247e09bee007894df8c081/kratos/includes/gid_io.h#L679

**Comment by armingeiser** [comment]
Already in the python process the closing of the FinalizeResults function is called only for multiple result files.

https://github.com/KratosMultiphysics/Kratos/blob/55e7608139955207e0247e09bee007894df8c081/kratos/python_scripts/gid_output_process.py#L287

I tried closing the file after each step also for binary mode, but then it gets overwritten all the time, and i only see the results of the last step. However, closing does solve the issue with the importing error in GiD.

I also tried to close and reopen immediately after. Then the import error with the unexpected end of file appears again.

Closing and appending seems to be a solution. But as i understand from the manual of gidpost, there is no "append" option for opening a file?
https://www.gidhome.com/documents/gidpost/Functions+references/Results+file+functions/GiD_fOpenPostResultFile
